File: HW1/Tests/InputParserTester.cs
using Xunit;
using static HW1.Calculator.InputParser;
using static HW1.Calculator.Calculator;

namespace Tests
{
    public class InputParserTester
    {
        private void Check(string[] args, int expectedCode)
        {
            var actualCode = TryParseArguments(args, out _, out _, out _);
            Assert.Equal(expectedCode, actualCode);
        }

        private void Check(string[] args, int expectedCode, int expVal1, Operation expOperation, int expVal2)
        {
            var actualCode = TryParseArguments(args, out var val1, out var operation, out var val2);
            Assert.Equal(expectedCode, actualCode);
            Assert.Equal(expVal1, val1);
            Assert.Equal(expOperation, operation);
            Assert.Equal(expVal2, val2);
        }

        [Fact]
        public void ArgsCountIsNotValid()
        {
            Check(new[] {"100", "/", "2", "10"}, WrongArgsCount);
        }

        [Fact]
        public void AreNotValidArgs()
        {
            Check(new[] {"x", "/", "2"}, WrongArgs);
            Check(new[] {"5", "*", "c"}, WrongArgs);
        }

        [Fact]
        public void IsNotValidOperationTest()
        {
            Check(new[] {"10", "x", "20"}, WrongOperation);
        }

        [Fact]
        public void DivisionByZeroError()
        {
            Check(new[] {"5", "/", "0"}, DivideByZero);
        }

        [Fact]
        public void AllIsCorrect()
        {
            Check(new[] {"5", "+", "7"}, 0, 5, Operation.Plus, 7);
            Check(new[] {"34", "-", "9"}, 0, 34, Operation.Minus, 9);
            Check(new[] {"44", "*", "39"}, 0, 44, Operation.Multiply, 39);
            Check(new[] {"54", "/", "71"}, 0, 54, Operation.Divide, 71);
        }
    }
}

File: HW4/Tests/InputParserTester.fs
module Tests.InputParserTester

open HW4
open Xunit
open ResultCodes

let check(args:string[]) (expectedCode:ResultCodes)=
    let mutable val1 = 0
    let mutable operation = ""
    let mutable val2 = 0
    let actualCode = InputParser.tryParseArguments args &val1 &operation &val2
    Assert.Equal(expectedCode, actualCode)

[<Fact>]
let ``WrongArgumentsCount``() =
    let args = [|"1"; "+"; "123"; "="|]
    check args ResultCodes.WrongArgumentsCount
    
let wrongArguments = [|
    [|[|"x"; "/"; "10"|]|]
    [|[|"5"; "*"; "x"|]|]
    [|[|"Hello"; "+"; "World"|]|]
    [|[|"Hello"; "-"; "156"|]|]
|]
[<Theory>]
[<MemberData(nameof(wrongArguments))>]
let ``TryParseArguments_WrongArguments``args =
    check args ResultCodes.WrongArguments

[<Fact>]
let ``TryParseArguments_WrongOperation``=
    let args = [|"10"; "?"; "12"|]
    check args ResultCodes.WrongOperation

let divideByZeroArguments = [|
    [|[|"144"; "/"; "0"|]|]
    [|[|"131"; ":"; "0"|]|]
|]

[<Theory>]
[<MemberData(nameof(divideByZeroArguments))>]
let ``TryParseArguments_DivideByZero``args =
    check args ResultCodes.DivideByZero
    
let correctArguments = [|
    [|[|"1"; "+"; "12"|]|]
    [|[|"13"; "-"; "56"|]|]
    [|[|"144"; "*"; "170"|]|]
    [|[|"7"; "/"; "41"|]|]
    [|[|"81"; ":"; "23"|]|]
|]

[<Theory>]
[<MemberData(nameof(correctArguments))>]
let ``TryParseArguments_AllCorrect`` args =
    check args ResultCodes.AllCorrect